src/Val.h

```diff
@@ -1518,7 +1518,9 @@ class RecordVal final : public Val, public notifier::detail::Modifiable
 	RecordTypePtr rt;
 
 	// Low-level values of each of the fields.
-	std::vector<std::optional<ZVal>>* record_val;
+	//
+	// Lazily modified during GetField(), so mutable.
+	mutable std::vector<std::optional<ZVal>> record_val;
```

Comment:
Just an observation and since you are already at it, `ZVal` can already encode an empty value (exactly what it's default ctor does), there shouldn't be a reason to additionally wrap it in an `optional` here (which adds another indirection at runtime). All you might need is to add an `explicit operator bool` to it, and adjust the usages from e.g., `*zval` to `zval`, and `zval->x` to `zval.x`.